### PR TITLE
Slaves

### PR DESCRIPTION
**Pull request by david-caro**
Added the slaves classes.

### CODE REVIEW

**Comment by david-caro** [comment]
Fixes #39

**Comment by david-caro** [comment]
@ekohl: about you two first comments, my bad, forgot to pull from rtylers fork before pushing, sorry.

About the keys, no, it's not a good idea, that's why those keys are only an example of what you should put in there and not meant to be used. Any improvement will be greatly appreciated (I'll try to improve it also, but I'm a little busy lately, and even if it's not the best solution it's one that works right now)

File: manifests/init.pp [hunk 1]
```diff
@@ -1,13 +1,54 @@
-class jenkins($version = 'installed') {
+
+class jenkins ($version='installed', $ui_user='', $ui_pass='') {
   include jenkins::repo
-  class {
-    'jenkins::package':
-      version => $version,
-  }
+  include jenkins::package
```
Comment (ekohl) [inline manifests/init.pp]: You should revert this back to `class {}` as well.

File: manifests/package.pp [hunk 2]
```diff
@@ -1,7 +1,9 @@
-class jenkins::package($version = 'installed') {
+class jenkins::package {
   package {
     'jenkins' :
-      ensure => $version;
```
Comment (ekohl) [inline manifests/package.pp:4]: And revert this part as well.

File: manifests/package.pp [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
+      ensure => installed;
+    'java-1.6.0-openjdk':
```
Comment (ekohl) [inline manifests/package.pp]: This is a very specific EL/Fedora package name and I'd recommend moving the package name to params.pp and determine it there.

File: manifests/plugin.pp [hunk 4]
```diff
@@ -1,13 +1,14 @@
-define jenkins::plugin($version=0) {
+define jenkins::plugin( $version=0 ) {
```
Comment (ekohl) [inline manifests/plugin.pp]: I'd leave this change out to keep the diff smaller, but not a big deal.

File: manifests/repo.pp [hunk 5]
```diff
@@ -2,23 +2,23 @@
```
Comment (ekohl) [inline manifests/repo.pp]: Could you revert the changes in this file?

File: manifests/slave.pp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      master => $master,
+      num_exec => $num_exec,
+      desc => $desc,
+      remote_fs => $remote_fs,
+  }
+
+  file {
+    $remote_fs:
+      ensure => directory,
+      mode   => '755',
```
Comment (ekohl) [inline manifests/slave.pp]: I think it's recommended to use 0755.

File: manifests/init.pp [hunk 7]
```diff
@@ -1,13 +1,54 @@
...
+  }
+
+  # Get the contents of the public key, note that 'file' adds a trailing newline
+  $key_path = split($settings::modulepath, ':')
+  $key_content = file("${key_path[0]}/jenkins/files/id_rsa.pub")
+  $key_content_fixed = split($key_content, '\n')
+
+  # declare this master's ssh key for the slaves to collect
+  @@ssh_authorized_key {
+    "$::hostname":
```
Comment (ekohl) [inline manifests/init.pp]: The quotes aren't needed.

**Comment by ekohl** [comment]
I'm really wondering what the best solution for the ssh key could be. This solution also makes it hard to replace it by users. Maybe one could require the user to define `file{'jenkins_master_ssh_private_key':}` and check if that's defined? My puppet knowledge is limited so I'm not sure what the best solution would be.

**Comment by david-caro** [comment]
I think that the best solution (not sure if it's possible) is to get the key from the master's jenkins user, or if not present, generate it, this way it does not get explicitly defined.

**Comment by ekohl** [comment]
Executing ssh-keygen should work. It also makes rotating quite easy because you simply remove the file and it will rotate.

**Comment by david-caro** [comment]
The problem is getting that ssh key to all the slaves, I have to investigate a little more on how to do that. Any insights are appreciated ;)

**Comment by ekohl** [comment]
I'm afraid I have no good suggestion here, but maybe you can write a custom provider or something?

**Comment by matthewbarr** [comment]
I'm also not thrilled with the slave addition code- it's very complicated to read.     We've just used a meta module, and set the node configuration up manually.

Maybe a Template of the main config.xml?    (Also, I've refactored some of the code to allow for more options in using repos & LTS.  If we can merge that, it'd be helpful.

**Comment by david-caro** [comment]
The main feature that I want is the automatic addition of slaves, without
human intervention. And if possible, without restarting the service.

To do that and help with the readability I implemented a custom type and a
new fact (actually, the code that uses them is quite simple). But I'm new
writing puppet types and providers, and also kind of new to ruby, so any
advice is appreciated.

The final usability of the module I think that is quite simple, and if you
need more control you can always use the type instead of the class, also
nice.

The ugly code is concentrated in the provider and fact implementations.

Using a template of the config.xml file will force you to reload the
service and you won't be able (as far as I know) to manage the slave is
online/offline status (I'm not sure if the slave will be connected the
first time neither :s).

About your refactor, I think that it's better not to mix commits, just wait
for the pull request to be accepted and I'll afterwards.

Btw. Thanks!
David

**Comment by matthewbarr** [comment]
Wasn't suggesting the inclusion of the repos / lts code into this - that'd be silly :)

One possibility for the slaves would be store config code, if you wanted, and the template. I guess it depends on how often you add slaves.

As for online / offline - slaves come online, assuming they are ssh, as soon as jenkins comes online.  Otherwise, they'd come online with the launch of the jar file.

I'm wondering if this isn't being solved on the wrong end.  Clearly this is a necessary function, to add slaves via programatic means.  Has the Jenkins world dealt with the issue?  

(Also, their RPM doesn't declare a dependency on java.  Very odd ;)

**Comment by matthewbarr** [comment]
Something like :  https://wiki.jenkins-ci.org/display/JENKINS/Swarm+Plugin or another design..?

**Comment by matthewbarr** [comment]
OK: did a quick test.  It's fairly trivial to use the swarm plugin.  Works very nicely, with a single exec on the slave side, and a single plugin on the master.    I already had to have a meta-module to create my slave environment and install the various packages we need, but i'll add a new commit / branch w/ the code to do this another way.  

(in short, follow the directions on the swarm plugin page.  )

**Comment by david-caro** [comment]
Hey Matthew,
My code does already use storecongifs (to be able to declare exported resources and collect them you need storeconfigs.

As far as I've seen there are 3 possible states for a slave:
- disconnected and offline
- connected and offline  <- this one will be missing if you only handle the config file and if present, will not be detected nor corrected.
- connected and online

About the swarm plugin, yes it does connect the slaves, but you have to control the execution and installation of another 'service' (the client jar file and do not forget the firewall rules, udp broadcast is usually restricted), and you have to install  the plugin on the master. Obviously you can do it that way, but I prefer not having to handle the execution of the jar file and installing the minimum plugins possible (and messing as less ass possible with the fw rules). Make it work out of the box when able.

The swarm plugin will be a great option if you do not have control of the created slaves, for example, if they do not run puppet or they are part of a big public project where people can 'donate' resources running a vm (that's what they built it for).

You also have the jenkinsapi python api to deal with automated slave creation, but right now it has some bug in the create_node function :S

btw. what's a meta-module for you? I've found that example42 has some meta-modules implementation, but not sure if you are referring to that, I wouldn't like to depend on all their modules though.

**Comment by matthewbarr** [comment]
Meta module is just what i'm calling the modules that actually call jenkins::plugin or jenkins.  I don't want to put all that in my node definition, so i've got a ayijenkins module that does things like RSA keys, installs the proper Repo for us, and has all the plugins.

Swarm doesn't require UDP, since you can just provide the URL of the master.  No more firewall than the code you're using, in fact. I got it working last night, with authentication, and no auto-discovery.  

I'll write up the module, and post the branch.  I think it makes life a bunch easier, but it does loose a tiny bit of tracking  of offline nodes, I suppose.

**Comment by david-caro** [comment]
And don't you think that all that you are doing with the 'meta-module' should be done by the jenkins module? So instead of having to write that extra module just passing the correct parameters should be enough?

In my opinion a module to install and manage jenkins should be capable of installing and managing jenkins... it seems to me that if you still need to write a wrapper just to handle that module then the module is not finished.

Did not know that it does not need udp, nice. 

How do you know if a slave is already connected? Having that process running does not mean that it's connected to the master, the process just retries to connect. Will the process run as a service? With init script and so on? (please, remember the return codes :S). And will it have it's own logfile? (the output of the jar is not logged in the server) If so, you should also add the logrotate entry. And the file will be downloaded from the plugin page directly or provided with the module? With the right version? (the same that you will need to have installed in the master, be careful if you have two masters with different versions or if you upgrade the plugin, the slave might not be updated).

I really do not like the idea of having to install an extra plugin to use a built-in jenkins feature...

File: files/id_rsa [hunk 8]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (rtyler) [inline files/id_rsa:1]: This doesn't seem like it should be in here :)

**Comment by rtyler** [comment]
I've not been paying attention the past couple of weeks, clearly! What's the status of this work? It's certainly a bit of work but I'm not sure with all the discussion whether I should merge and release it at this point or not. What do you think @david-caro

**Comment by matthewbarr** [comment]
(i have a longer response at the office, as i got busy working on something else, and didn't get to commit & finish the entire parameterization effort for slaves.  There's a _ton_ of parameters to deal w/ a user's need to customize.)

**Comment by david-caro** [comment]
@rtyler: I'm ok with my proposal, but Matthew seems to differ and he wanted
to propose a different solution, maybe we should wait and see which one is
better, your call.

**Comment by david-caro** [comment]
@matthewbarr: How are you doing? Any progress?

**Comment by matthewbarr** [comment]
Yep! go look at the PR for it:

https://github.com/rtyler/puppet-jenkins/pull/41

I got things to do what I wanted finally, and added what I think are enough parameters for everyone.  Make sure to look at the docs, to see how to use it.    Fairly easy, but you do need to do something to make it work.

**Comment by david-caro** [comment]
It seems that this is not needed anymore. As Matthew pull got merged. Abandoning it.
